lib/CPathfinder.cpp

```diff
@@ -213,60 +228,120 @@ void CPathfinder::addTeleportExits(bool noTeleportExcludes)
 	}
 }
 
-bool CPathfinder::isMovementPossible()
+bool CPathfinder::isLayerTransitionPossible()
 {
-	if(!canMoveBetween(cp->coord, dp->coord) || dp->accessible == CGPathNode::BLOCKED)
+	if((cp->layer == EPathfindingLayer::AIR || cp->layer ==  EPathfindingLayer::WATER)
+	   && dp->layer != EPathfindingLayer::LAND)
+	{
+		return false;
+	}
+	else if(cp->layer == EPathfindingLayer::LAND && dp->layer == EPathfindingLayer::AIR)
+	{
+		if(options.lightweightFlyingMode && !isSourceInitialPosition())
+			return false;
+	}
+	else if(cp->layer == EPathfindingLayer::SAIL && dp->layer != EPathfindingLayer::LAND)
 		return false;
+	else if(cp->layer == EPathfindingLayer::SAIL && dp->layer == EPathfindingLayer::LAND)
+	{
+		if(!dt->isCoastal())
+			return false;
 
-	Obj destTopVisObjID = dt->topVisitableId();
-	if(cp->land != dp->land) //hero can traverse land<->sea only in special circumstances
+		//tile must be accessible -> exception: unblocked blockvis tiles -> clear but guarded by nearby monster coast
+		if((dp->accessible != CGPathNode::ACCESSIBLE && (dp->accessible != CGPathNode::BLOCKVIS || dt->blocked))
+			|| dt->visitable)  //TODO: passableness problem -> town says it's passable (thus accessible) but we obviously can't disembark onto town gate
+			return false;
+
+		useEmbarkCost = 2;
```

Comment:
I'm going to make it constant once I find out what to do with cost calculations.